mPower2/mPower2/PassiveCollectors.swift

```diff
@@ -0,0 +1,291 @@
+        dataPoint[kVerticalAccuracyKey] = verticalAccuracy
+        
+        return dataPoint
+    }
+    
+    func uploadDisplacement(_ displacementData: [String: RSDJSONValue]) {
+        let archiveFilename = kPassiveDisplacementArchiveFilename
+        let archive = SBBDataArchive(reference: schemaIdentifier, jsonValidationMapping: nil)
+        if let schemaRevision = MP2BridgeConfiguration.shared.schemaInfo(for: schemaIdentifier)?.schemaVersion {
+            archive.setArchiveInfoObject(schemaRevision, forKey: kSchemaRevisionKey)
+        }
+        archive.insertDictionary(intoArchive: displacementData, filename: archiveFilename, createdOn: Date())
+        
+        do {
+            try archive.complete()
+            archive.encryptAndUploadArchive()
+        }
+        catch {}
+   }
+    
+    // MARK: CLLocationManagerDelegate
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        if status == .authorizedAlways {
+            // hey we have permission now, yay! so (re)start monitoring
+            start()
+        } else {
+            // we no longer have permission, but we can still upload what we collected while we did
+            stop(discarding: false)
+        }
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        // Generate & upload (a) displacement(s) from the given location(s)
+        let validLocations = locations.filter({$0.horizontalAccuracy >= 0.0})
+        for location in validLocations {
+            if let lastLocation = previousLocation {
+                let dataPoint = displacementData(for: location, previousLocation: lastLocation)
+                uploadDisplacement(dataPoint)
+            }
+            previousLocation = location
+        }
+    }
+}
+
+/// This class allows us (with participant permission) to collect gait information passively,
+/// in the background, as they go about their day.
+class PassiveGaitCollector : PassiveCollector {
+    /// The schema identifier to which the collector should upload the data it collects.
+    var schemaIdentifier: String
+
+    /// Create an instance with the specified schema identifier.
+    init(with identifier: String) {
+        schemaIdentifier = identifier
+    }
+    
+    /// Start the passive gait collector.
+    func start() {
+        
+    }
+    
+    /// Stop the passive gait collector.
+    /// - parameter discarding: true if pending gait data since the last upload should be discarded, false if it should be uploaded.
+    func stop(discarding: Bool) {
+        
+    }
+}
+
+
+extension CLLocation {
+    static func toRadians(from degrees: CLLocationDegrees) -> Double {
+        return (degrees / 180.0) * .pi
+    }
+    
+    static func toDegrees(from radians: Double) -> CLLocationDegrees {
+        return (radians / .pi) * 180.0
+    }
+    
+    func bearingInRadians(to endLocation: CLLocation) -> Double {
```

Comment:
I think I remember checking and finding out that `CLLocation.course` is the same thing.

Reply:
`CLLocation.course` is actually the direction in which the device is currently traveling, not the bearing from the previous location reading to the current one (which is what we want), and also--and more importantly--is only available if GPS is active, which it would only be if they happen to be using GPS for something else (like Maps) at the time it gets the significant location update.

Reply:
(also course is in degrees and they wanted radians because it's easier for calculations, but admittedly that's a trivial conversion.)